File: prototype/node/src/app/core/analytics/tally.ts
/** How many rows carry one key — a status, an event type, a ledger entry type. */
export type Tally<Key extends string> = { key: Key; count: number }

/**
 * Every key the domain knows, in the domain's own order, carrying the count
 * that was measured for it. A `group by` answers only for the keys that have
 * rows, and a status nobody has reached is still a status the page shows.
 */
export function tallyOver<Key extends string>(
  keys: readonly Key[],
  counted: readonly Tally<Key>[],
): readonly Tally<Key>[] {
  const counts = new Map(counted.map((tally) => [tally.key, tally.count]))

  return keys.map((key) => ({ key, count: counts.get(key) ?? 0 }))
}

export function tallyTotal<Key extends string>(tallies: readonly Tally<Key>[]): number {
  return tallies.reduce((total, tally) => total + tally.count, 0)
}

File: prototype/rails/src/app/models/tally.rb
# Every key a domain enum names, in the enum's own order, paired with the
# count a group-by measured for it — 0 for a key nothing has reached. A
# `group by` answers only for the keys that have rows, and a status nobody
# has reached is still a status the dashboard shows.
module Tally
  def self.over(keys, counted)
    keys.index_with { |key| counted.fetch(key.to_s, 0) }
  end
end
